bob/src/cluster/quorum.rs

```diff
@@ -294,3 +340,73 @@ impl Cluster for Quorum {
         self.delete_on_nodes(key).await
     }
 }
+
+fn filter_local_keys(keys: &[BobKey], mapper: &Virtual) -> (Vec<usize>, Vec<BobKey>)  {
+    let mut indices = Vec::new();
+    let local_keys = keys
+        .iter()
+        .enumerate()
+        .filter_map(|(idx, &key)| {
+            mapper.get_target_nodes_for_key(key).iter().find_map(|node| {
+              if node.name() == mapper.local_node_name() {
+                  indices.push(idx);
+                  Some(key)
+              } else {
+                  None
+              }
+            })
+        })
+        .collect::<Vec<BobKey>>();
+    (indices, local_keys)
+}
+
+fn filter_not_found(exist: &[bool], keys: &[BobKey]) -> Vec<BobKey> {
+    let mut not_found_keys = Vec::new();
+    for (idx, &r) in exist.iter().enumerate() {
+        if r == false {
+            not_found_keys.push(keys[idx]);
+        }
+    }
+    not_found_keys
+}
+
+fn filter_remote_not_found_by_nodes(exist: &[bool], keys: &[BobKey], mapper: &Virtual) -> HashMap<Vec<Node>, (Vec<BobKey>, Vec<usize>)> {
```

Comment:
I don't like grouping by `Vec<Node>`: this is a strong indicator of poor program logic